### PR TITLE
lex identifiers faster

### PR DESCRIPTION
**Pull request by KristofferC**
I thought we could first put the faster identifier in master and then we can think a bit more about the stringless tokens.

### CODE REVIEW

File: src/lexer.jl [hunk 1]
```diff
@@ -312,7 +316,7 @@ function next_token(l::Lexer)
     elseif c == 'i'; return lex_i(l);
```
Comment (ZacLN) [inline src/lexer.jl]: this can be removed

File: src/lexer.jl [hunk 2]
```diff
@@ -312,7 +316,7 @@ function next_token(l::Lexer)
...
     elseif c == 't' || c == 'f'; return lex_bool(l);
```
Comment (ZacLN) [inline src/lexer.jl]: as can this

Comment (ZacLN) [reply src/lexer.jl]: and their respective functions

Comment (KristofferC) [reply src/lexer.jl]: The `"1 isa 2"` test fails on 0.5 then.

Comment (KristofferC) [reply src/lexer.jl]: Maybe the new identifier isn't set up to work with that?

Comment (KristofferC) [reply src/lexer.jl]: I pushed it anyway for now

Comment (ZacLN) [reply src/lexer.jl]: Ah ok, will a ` && VERSION >= v"0.6.0-dev.1471"`on line 965 of `lexer.jl`  fix it?

Comment (KristofferC) [reply src/lexer.jl]: yes

**Comment by ZacLN** [comment]
quotes on c!

**Comment by KristofferC** [comment]
Wops! Fixed

**Comment by ZacLN** [comment]
This has uncaught errors don't merge!

File: src/lexer.jl [hunk 3]
```diff
@@ -731,4 +733,348 @@ function lex_cmd(l::Lexer)
     end
 end
 
+function tryread(l, str, k)
+    for s in str
+        c = readchar(l)
+        if c!=s
+            if !is_identifier_char(c)
+                skip(l.io, -Int(!eof(l.io)))
```
Comment (ZacLN) [inline src/lexer.jl]: replace with `backup!(l)

File: src/lexer.jl [hunk 4]
```diff
@@ -731,4 +733,348 @@ function lex_cmd(l::Lexer)
...
+    elseif c == 'e'
+        c = readchar(l)
+        if c == 'l'
+            c = readchar(l)
+            if c == 's'
+                c = readchar(l)
+                if c == 'e'
+                    c = readchar(l)
+                    if !is_identifier_char(c)
+                        skip(l.io, -Int(!eof(l.io)))
```
Comment (ZacLN) [inline src/lexer.jl]: replace with `backup!(l)`

File: src/lexer.jl [hunk 5]
```diff
@@ -731,4 +733,348 @@ function lex_cmd(l::Lexer)
...
+            elseif c == 'p'
+                c = readchar(l)
+                if c == 'o'
+                    c = readchar(l)
+                    if c == 'r'
+                        c = readchar(l)
+                        if c == 't'
+                            c = readchar(l)
+                            if !is_identifier_char(c)
+                                skip(l.io, -Int(!eof(c)))
```
Comment (ZacLN) [inline src/lexer.jl]: replace with `backup!(l)`

File: src/lexer.jl [hunk 6]
```diff
@@ -731,4 +733,348 @@ function lex_cmd(l::Lexer)
...
+                end
+            end
+        elseif c == 'y'
+            c = readchar(l)
+            if c == 'p'
+                c = readchar(l)
+                if c == 'e'
+                    c = readchar(l)
+                    if !is_identifier_char(c)
+                        skip(l.io, -Int(!eof(l.io)))
```
Comment (ZacLN) [inline src/lexer.jl]: replace with `backup!(l)`

**Comment by ZacLN** [comment]
a blanket replacement of `skip(l.io, -Int(!eof(l.io)))` with `backup!(l)` is needed

**Comment by KristofferC** [comment]
Could you share the tests that you used to find the problems?

**Comment by ZacLN** [comment]
It came up from the parser:
```@test length(collect(tokenize("x)")))==3```
Before it was incorrectly reading it as:
```
IDENTIFIER    "x)"
ENDMARKER     ""
```

**Comment by KristofferC** [comment]
Updated with some tests and refactored out a common chunk of code into a function.

**Comment by KristofferC** [comment]
Whenever you have time, perhaps you can try this PR with your Parser and see if things still work ok and the speedup is retained.

**Comment by ZacLN** [comment]
Yep it's up and working, the only fiddle was the change in how byte positions are stored

File: src/lexer.jl [hunk 7]
```diff
@@ -731,4 +733,251 @@ function lex_cmd(l::Lexer)
...
+function _doret(c, l)
+    if !is_identifier_char(c)
+        backup!(l)
+        return emit(l, IDENTIFIER)
+    else
+        return readrest(l)
+    end
+end
+
+function lex_identifier(l, c)
```
Comment (tkelman) [inline src/lexer.jl:770]: this strikes me as something that should maybe be autogenerated

Comment (KristofferC) [reply src/lexer.jl:770]: Oh yes, it definitely should.

Comment (ZacLN) [reply src/lexer.jl:770]: Alright then, I'll have a look at this tomorrow

Comment (tkelman) [reply src/lexer.jl:770]: wonder if Automa.jl would be good for this kind of thing

**Pull request merged**